dotCMS/src/main/java/com/dotcms/auth/providers/saml/v1/SAMLHelper.java

```diff
@@ -0,0 +1,468 @@
+
+                            Logger.debug(this, () -> "Skipping role: " + role);
+                            continue;
+                        } else {
+
+                            Logger.debug(this, () -> "Role Patterns: " + this.toString(rolePatterns) + ", remove role prefix: "
+                                    + removeRolePrefix + ": true");
+                        }
+                    }
+
+                    this.addRole(user, removeRolePrefix, role);
+                }
+            }
+
+            return;
+        }
+
+        Logger.info(this, "Roles have been ignore by the build roles strategy: " + buildRolesStrategy
+                + ", or roles have been not set from the IdP");
+    }
+
+    private void addRole(final User user, final String removeRolePrefix, final String roleObject)
+            throws DotDataException {
+
+        // remove role prefix
+        final String roleKey = isSet(removeRolePrefix)?
+                roleObject.replaceFirst(removeRolePrefix, StringUtils.EMPTY):
+                roleObject;
+
+        addRole(user, roleKey, false, false);
+    }
+
+    private void addRole(final User user, final String roleKey, final boolean createRole, final boolean isSystem)
+            throws DotDataException {
+
+        Role role = this.roleAPI.loadRoleByKey(roleKey);
+
+        // create the role, in case it does not exist
+        if (role == null && createRole) {
+            Logger.info(this, "Role with key '" + roleKey + "' was not found. Creating it...");
+            role = createNewRole(roleKey, isSystem);
+        }
+
+        if (null != role) {
+            if (!this.roleAPI.doesUserHaveRole(user, role)) {
+                this.roleAPI.addRoleToUser(role, user);
+                Logger.debug(this, "Role named '" + role.getName() + "' has been added to user: " + user.getEmailAddress());
+            } else {
+                Logger.debug(this,
+                        "User '" + user.getEmailAddress() + "' already has the role '" + role + "'. Skipping assignment...");
+            }
+        } else {
+            Logger.debug(this, "Role named '" + roleKey + "' does NOT exists in dotCMS. Ignoring it...");
+        }
+    }
+
+    private Role createNewRole(String roleKey, boolean isSystem) throws DotDataException {
+        Role role = new Role();
+        role.setName(roleKey);
+        role.setRoleKey(roleKey);
+        role.setEditUsers(true);
+        role.setEditPermissions(false);
+        role.setEditLayouts(false);
+        role.setDescription("");
+        role.setId(UUIDGenerator.generateUuid());
+
+        // Setting SYSTEM role as a parent
+        role.setSystem(isSystem);
+        Role parentRole = roleAPI.loadRoleByKey(Role.SYSTEM);
+        role.setParent(parentRole.getId());
+
+        String date = DateUtil.getCurrentDate();
+
+        ActivityLogger.logInfo(ActivityLogger.class, getClass() + " - Adding Role",
+                "Date: " + date + "; " + "Role:" + roleKey);
```

Comment:
Codacy found an issue: [The String literal "Date: " appears 4 times in this file; the first occurrence is on line 350](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5707914)